File: python/Q567_permutation_in_string.py
#
# @lc app=leetcode id=567 lang=python3
#
# [567] Permutation in String
#


# @lc code=start
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        s1_dict: dict[str, int] = dict()
        for s in s1:
            s1_dict[s] = s1_dict.get(s, 0) + 1

        left: int = 0
        remained: int = len(s1_dict)
        counter: dict[str, int] = dict()

        for right in range(len(s2)):
            letter: str = s2[right]

            if letter not in s1_dict:
                remained = len(s1_dict)
                counter.clear()
                left = right + 1
                continue

            counter[letter] = counter.get(letter, 0) + 1

            if letter in s1_dict and counter[letter] == s1_dict[letter]:
                remained -= 1

            while left <= right and counter[letter] > s1_dict[letter]:
                l_letter: str = s2[left]
                if l_letter in s1_dict and counter[l_letter] == s1_dict[l_letter]:
                    remained += 1

                counter[l_letter] -= 1
                left += 1

            if remained == 0:
                return True

        return False


# @lc code=end

File: rust/src/leetcode/Q567_permutation_in_string.rs
/*
 * @lc app=leetcode id=567 lang=rust
 *
 * [567] Permutation in String
 */

// @lc code=start
use std::collections::HashMap;

impl Solution {
    pub fn check_inclusion(s1: String, s2: String) -> bool {
        let s1_vec: Vec<char> = s1.chars().collect();
        let mut s1_map: HashMap<char, i32> = HashMap::new();
        for c in s1_vec {
            *s1_map.entry(c).or_insert(0) += 1;
        }

        let s2_vec: Vec<char> = s2.chars().collect();
        let mut s2_map: HashMap<char, i32> = HashMap::new();

        let mut left: usize = 0;
        let mut remained: usize = s1_map.len();

        for right in 0..s2_vec.len() {
            let right_char: char = s2_vec[right];

            if !s1_map.contains_key(&right_char) {
                remained = s1_map.len();
                s2_map.clear();
                left = right + 1;
                continue;
            }

            *s2_map.entry(right_char).or_insert(0) += 1;

            if s1_map.get(&right_char) == s2_map.get(&right_char) {
                remained -= 1;
                if remained == 0 {
                    return true;
                }
            } else {
                while s1_map.get(&right_char) < s2_map.get(&right_char) {
                    let left_char: char = s2_vec[left];

                    if s1_map.get(&left_char) == s2_map.get(&left_char) {
                        remained += 1;
                    }

                    *s2_map.get_mut(&left_char).unwrap() -= 1;
                    left += 1;
                }
            }
        }

        false
    }
}
// @lc code=end
